app/serializers/alchemy/page_tree_serializer.rb

```diff
@@ -56,6 +56,7 @@ def page_hash(page, has_children, level, folded)
         visible: page.visible?,
         restricted: page.restricted?,
         page_layout: page.page_layout,
+        allowing_sub_pages: page.definition['sub_pages'] != false,
```

Comment:
You probably want to ask `.present?` instead here, to account for page definitions that do not have `sub_pages` set.

Reply:
> You probably want to ask .present? instead here, to account for page definitions that do not have sub_pages set.

This feature was intended as an optional "restriction" rather than required specification. If they don't have it set I think all sub pages should be available as normal by default. Only if explicitly set it would restrict availability, and only if set to false should the button be removed.

> I have a few implementation comments, and I would like to suggest to only implement ONE of `sub_pages` or `parent_pages`. If a page layout can only have certain subpages, then that's all you should ever need, am I right?

I think it's worth it to be able to say that `blog_post` is only available under `blog` so I don't have to list all other layouts except `blog_post` on all other layouts.

Another thought I had now is that this previously used page_layout property `layout['hide']` (mentioned in `allowed_in_definition`), could be replaced by `parent_pages = false`. Meaning that no pages allow creating it => it's hidden. What do you think?

Reply:
I get it, ok. I'm still not happy with the idea of having the page_layouts array for one page having to check the definitions of all other page layouts. 

What about having an `only` and an `except` key in the page_layouts.yml: 
```yml
# config/alchemy/page_layouts.yml

- name: home
  sub_pages: 
    except: 
    - blog_post # array of allowed sub_pages.

- name: blog
  sub_pages: 
    only: 
    - blog_post # array of allowed sub_pages.

- name: blog_post
  sub_pages:
    only: []
``

Reply:
If you look at `page_layout.rb` before my changes (below). It still needs to loop through each layout to check for uniqueness and "hide" etc. So we're already doing these cross layout conditions. And i think it is actually really convenient/useful to have the `parent_page` restriction as well so you don't have to check and repeat what to exclude on every page. I doesn't seem to cost anything to add another condition in there.
```
def selectable_layouts(language_id, only_layoutpages = false)
  @language_id = language_id
  all.select do |layout|
    if only_layoutpages
      layout['layoutpage'] && layout_available?(layout)
    else
      !layout['layoutpage'] && layout_available?(layout)
    end
  end
end

def layout_available?(layout)
  !layout['hide'] && !already_taken?(layout) && available_on_site?(layout)
end
```

But I still like your idea of exclude/include. What about this pattern (below) that would give the user all the options? (`*` = all, `!` = not)

Then we could say that the default value of `sub_pages`/`parent_pages` is `[*]` unless specified. And we could replace this `layout['hide']`/`hide: true`  option with `parent_pages: false`

```
- name: blog
  sub_pages: [blog_post] # array of allowed sub_pages.

- name: blog_post
  parent_pages: [blog] # array of parent pages under which page can be created
  sub_pages: false


- name: default
  parent_pages: [*]
  sub_pages: [*] 

- name: all_except_blog_as_sub_page
  sub_pages: [*, !blog]

- name: creatable_under_all_except_blog
  parent_pages: [*, !blog]

- name: hidden
  parent_pages: false
  # replaces hide: true
```

What do you think? Thanks

Reply:
@mamhoff, i don't know why i started thinking about this `*` and `!` syntax. I think your version with `only`/`except` is better :)

But do you think it's ok to keep the `parent_page` option? thanks